### PR TITLE
Refactor AnimationState

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: cocos-creator/3d-tasks#

Changelog:
 * Refactor `AnimationState`.

This PR investigates and refactor class `AnimationState`. This PR targets v3.1.

### CODE REVIEW

**Comment by shrinktofit** [comment]
### Change Details

- Introduces a new `Animation.EventType.ITERATION_END` enumerator which substitutes and deprecates `Animation.EventType.LASTFRAME`. The reason is `LASTFRAME` is misspelled and does not match its meaning.

- Supply details and missing API docs for `AnimationState`.

- Add unit tests for `AnimationState`.

- To decouple `AnimationState` from `AnimationManager`, the interface `AnimationState.Context` is introduced. This way also allow editor to customize when and how the animation states are updated. The context object:

    - Enqueues and dequeues animation states from update queue.

    - Emit playback events and frame events.

    - Specifies the blend state buffer that used to blend skeletal animations.

> NOTE: Verification is needed to ensure the following are not breaking changes @2youyou2 

- The "perfect first frame" mechanism is officially documented. "Perfect first frame" means that manually assigned time and initial time ignore the delta time increment at next `update()`, except for `sample()` is performed before `update()`.

- When the `time` arrives just at the animation's end, the `current` is set as `0` if next iteration's direction is same as current iteration. This does not hold if `time` equals to `(repeatCount * duration)`.

- `sample()` does not trigger any event, but `update()` does.

File: cocos/core/animation/animation-state.ts [hunk 1]
```diff
@@ -329,30 +357,64 @@ export class AnimationState extends Playable {
...
+    private _isIfl = false;
     private _delay = 0;
     private _delayTime = 0;
+    private _name: string;
+    private _currentPosition = new PlaybackPosition();
     /**
-     * Mark whether the current frame is played.
-     * When set new time to animation state, we should ensure the frame at the specified time being played at next update.
+     * Whether current time has been sampled.
+     * If not, we won't take the next time update through `update()` call.
```
Comment (shrinktofit) [inline cocos/core/animation/animation-state.ts:370]: Marks the `First perfect frame`

File: cocos/core/animation/animation-state.ts [hunk 2]
```diff
@@ -329,30 +357,64 @@ export class AnimationState extends Playable {
...
+     * Creates an animation state object that is used to play the specified animation clip.
+     * @param context The animation state's context.
+     * @param clip The associated animation clip.
+     * @param name Name of the animation state. If not specified or is empty, the clip's name would be taken.
+     */
+    constructor (context: AnimationState.Context, clip: AnimationClip, name?: string);
+
+    /**
+     * @zh
+     * 创建一个用于播放动画剪辑的动画状态对象。将使用全局上下文。
```
Comment (shrinktofit) [inline cocos/core/animation/animation-state.ts:394]: The global context is the animation manager stuff.

File: cocos/core/animation/animation-state.ts [hunk 3]
```diff
@@ -488,149 +552,98 @@ export class AnimationState extends Playable {
...
+     * 要避免完美首帧，需要确保更新后的时间已被应用。例如，通过 `sample()` 调用。
+     * @en Steps the playback time and do sampling.
+     * Note, this method may be affected by the "perfect first frame" mechanism:
+     * when the animation state is in initial state or its time got explicitly set through `this.time`
+     * and never been played(thought whatever `update()` or `sample()`),
+     * the next `update()` call would not change the play time but only apply then time.
+     * The reason for this is to prevent the first frame from being suddenly skipped as the animation came into view.
+     * To avoid the perfect first frame,
+     * ensure the updated time has been applied. For example, through `sample()` call.
+     * @param delta The delta time, in seconds.
```
Comment (shrinktofit) [inline cocos/core/animation/animation-state.ts:583]: Documents the "perfect first frame"

File: cocos/core/animation/animation-state.ts [hunk 4]
```diff
@@ -974,11 +891,158 @@ function isTargetingTRS (path: TargetPath[]) {
...
+        // If direction is inverse, increase index.
+        if (!forwarding) {
+            eventIndex += 1;
+        }
+    }
+    return eventIndex;
+}
+
+export declare namespace AnimationState {
+    export interface Context {
```
Comment (shrinktofit) [inline cocos/core/animation/animation-state.ts:915]: @gameall3d Take a look at here please. You're the main user.

**Comment by minggo** [comment]
@shrinktofit can close this PR?

**Comment by SantyWang** [comment]
Seems no progress. I close it now, and you can reopen it when it's ready

**Pull request closed without merging**